File: neetcode150/python-solutions/2_pointers/container_with_most_water.py
from typing import List


def max_area(height: List[int]):
    result: int = 0

    start: int = 0
    end: int = len(height) - 1

    while start < end:
        curr_area = (end - start) * min(height[start], height[end])
        result = max(curr_area, result)

        if height[end] > height[start]:
            start += 1
        else:
            end -= 1

    return result

File: neetcode150/rust-solutions/src/two_pointers/container_with_most_water.rs
use std::cmp::{max, min};

fn max_area(height: Vec<i32>) -> i32 {
    let mut left = 0;
    let mut right = height.len() - 1;
    let mut result = 0;


    while left < right {
        let area = min(height[left], height[right]) * (right - left) as i32;
        result = max(result, area as i32);
        if left > right {
            right -= 1;
        } else {
            left += 1;
        }
    }

    return result;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_max_area() {
        let height = vec![1, 8, 6, 2, 5, 4, 8, 3, 7];       
        let want = 49;
        let got = max_area(height);
        assert_eq!(want, got);
    }
}

